src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs

```diff
@@ -56,7 +57,42 @@ public void IdentifyBestMatchAndFilterToAllItems(CompletionFilterReason filterRe
             {
                 using (Logger.LogBlock(FunctionId.Completion_ModelComputation_FilterModelInBackground, CancellationToken.None))
                 {
-                    return FilterModelInBackgroundWorker(model, id, caretPosition, recheckCaretPosition, dismissIfEmptyAllowed, filterReason);
+                    if (models == default(ImmutableArray<Model>))
+                    {
+                        return default(ImmutableArray<Model>);
+                    }
+                    else
+                    {
+                        // We want to dismiss the session if the caret ever moved outside our bounds.
+                        if (recheckCaretPosition && Controller.IsCaretOutsideAllItemBounds(models, caretPosition))
+                        {
+                            return default(ImmutableArray<Model>);
+                        }
+
+                        var multipleModels = models.Length > 1;
+                        var filteredModels = models.Select(m => FilterModelInBackgroundWorker(m, id, caretPosition, recheckCaretPosition, dismissIfEmptyAllowed, filterReason, multipleModels))
+                                                    .ToImmutableArray();
+
+                        // Every model can truly no longer present anything.
+                        if (filteredModels.All(f => f == null))
+                        {
+                            return default(ImmutableArray<Model>);
+                        }
+
+                        // If the selected model doesn't have a selected item, try to choose one that does
+                        var selectedModel = filteredModels.FirstOrDefault(m => m.IsSelected);
+                        
+                        if (selectedModel != null && (selectedModel.SelectedItem == null || selectedModel.IsSoftSelection))
+                        {
+                            var nextSelectableModel = filteredModels.FirstOrDefault(m => m.SelectedItem != null && m.IsHardSelection);
+                            if (nextSelectableModel != null)
```

Comment:
For controlling the selection of a new model when the currently selected one has no filter matches.